### ISSUE TITLE
Charging time remaining incorrect value

### ISSUE THREAD

**Issue by rhoddan** [bimmerconnected/ha_custom_component#12]
![ECD29324-EE31-4ED5-9C8B-7EBBDEC52EDD](https://user-images.githubusercontent.com/33499995/142942633-adefb2ce-94e3-43ce-bfd7-02c6332877cf.jpeg)

The charging time remaining seems to be incorrect. I guess max 45 minutes left

**Comment by askthomas** [comment]
it in minutes, but hows in hours

13.73 hours should be 137.3 minutes

**Comment by rikroe** [comment]
Just fixed AM/PM parsing coming of the charging end times, please try again with the newest custom component/library.

**Comment by rhoddan** [comment]
20211127.1 Still mismatch between BMW App and HA regarding end charging time remaining...

![IMG_0155](https://user-images.githubusercontent.com/33499995/143690602-72375a94-18ce-413b-ab6b-0f760cd25bef.PNG)
![IMG_0156](https://user-images.githubusercontent.com/33499995/143690604-b878dfb5-0c3c-448c-bc98-88a9acba450e.jpeg)

**Comment by rikroe** [comment]
I was thinking of maybe removing `charging_time_remaining` completely and just replacing it with `charging_end_time`.
That could be easier to debug and produce less state changes in Home Assistant.
If needed you still could calculate the remaining time using a simple sensor in HA.

**Comment by rikroe** [comment]
Please check the new sensor `charging_end_time` in the latest custom component. If you want to see the value returned by the BMW API, please also activate the `charging_end_time_mybmw` sensor from device settings in HA.

**Comment by rhoddan** [comment]
![IMG_0158](https://user-images.githubusercontent.com/33499995/144767312-24fb0a01-ea8f-4315-9bc4-c5597375b37c.jpeg)
2021-12-05 22:39:23 ERROR (SyncWorker_2) [bimmer_connected.account] 401 Client Error: Unauthorized for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1638736763.002946&tireGuardMode=ENABLED
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py", line 300, in send_request
    response.raise_for_status()
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 953, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1638736763.002946&tireGuardMode=ENABLED
2021-12-05 22:39:23 ERROR (SyncWorker_2) [custom_components.bmw_connected_drive] Could not connect to the BMW Connected Drive portal. The vehicle state could not be updated
2021-12-05 22:39:23 ERROR (SyncWorker_2) [custom_components.bmw_connected_drive] 401 Client Error: Unauthorized for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1638736763.002946&tireGuardMode=ENABLED
Traceback (most recent call last):
  File "/config/custom_components/bmw_connected_drive/__init__.py", line 321, in update
    self.account.update_vehicle_states()
  File "/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py", line 412, in update_vehicle_states
    self._get_vehicles()
  File "/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py", line 376, in _get_vehicles
    response = self.send_request(
  File "/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py", line 303, in send_request
    raise ex
  File "/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py", line 300, in send_request
    response.raise_for_status()
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 953, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://cocoapi.bmwgroup.com/eadrax-vcs/v1/vehicles?apptimezone=60.0&appDateTime=1638736763.002946&tireGuardMode=ENABLED

**Comment by rikroe** [comment]
Sorry, you were not able to. Check your entities list with the new custom component: https://github.com/bimmerconnected/ha_custom_component/releases/tag/20211207.1

**Comment by rhoddan** [comment]
![IMG_0824](https://user-images.githubusercontent.com/33499995/145463560-ed1e4bba-e5bd-4543-897d-de1654a2d0a7.jpeg)
![Skärmavbild 2021-12-09 kl  20 32 33](https://user-images.githubusercontent.com/33499995/145463588-3febe0a3-2ee3-45a3-9a57-460f391e19f4.png)

Why does it differ between HA 03:06AM) (updated via bmw_connected_drive.update_state service) and my My BMW App (23:20)? It's also 12h & 24h time format

**Comment by rikroe** [comment]
This is driving me nuts. I really think we're using the wrong endpoint here...

**Comment by rikroe** [comment]
Can you please do me a favor and edit `/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py` [L384](https://github.com/bimmerconnected/bimmer_connected/blob/0.8.6/bimmer_connected/account.py#L384) to the following:
```python
                    "appDateTime": int(datetime.datetime.now().timestamp()*1000),
```
and test again?

The BMW API will always return 12h format, so we have to parse it.

If your `charging_end_time` sensor (not the `_original` one) shows *Invalid timestamp* (like in your screenshot above), can you please check the entity itself & if there are any logs why this is not working?

**Comment by rhoddan** [comment]
> Can you please do me a favor and edit `/usr/local/lib/python3.9/site-packages/bimmer_connected/account.py` [L384](https://github.com/bimmerconnected/bimmer_connected/blob/0.8.6/bimmer_connected/account.py#L384) to the following:
> 
> ```python
>                     "appDateTime": int(datetime.datetime.now().timestamp()*1000),
> ```
> 
> and test again?
> 
> The BMW API will always return 12h format, so we have to parse it.
> 
> If your `charging_end_time` sensor (not the `_original` one) shows _Invalid timestamp_ (like in your screenshot above), can you please check the entity itself & if there are any logs why this is not working?
![Skärmavbild 2021-12-10 kl  00 02 36](https://user-images.githubusercontent.com/33499995/145489454-03bfc20f-ac77-4a74-817a-f36c8bbe6c43.png)

I'm running Home Assistant "Supervised" and not sure how to edit that file? The only files I can edit is as the attached.

**Comment by rikroe** [comment]
Can you please edit `manifest.json` and change the line with `requirements` to: 
```json
  "requirements": ["git+https://github.com/bimmerconnected/bimmer_connected.git@master"],
```
That should bring you up to the latest and just merged master branch.

**Comment by rhoddan** [comment]
> Can you please edit `manifest.json` and change the line with `requirements` to:
> 
> ```json
>   "requirements": ["bimmer_connected==0.8.7.0b1"],
> ```
![Skärmavbild 2021-12-10 kl  23 28 07](https://user-images.githubusercontent.com/33499995/145649578-b953291b-5a37-468a-bea7-3248d3f1d6cf.png)

> 
> That should bring you up to the latest and just merged master branch.
> 
> EDIT: Updated to a pre-release version, as the `git+` link doesn't seem to work.

Thanks for all help and time. I edited the manifest.json (see attached) and then restarted HA I also tried to redownload but no new version available. Probably I'm missing a step?

**Comment by rikroe** [comment]
There is no new component available, so you should not see any upgrade notice in HA.

Please check with `pip3 freeze` if your are on the beta mentioned above and then have a look at the `charging_end_time` sensors and their values.

**Comment by rikroe** [comment]
Just looked through the PRs for HA 2021.12. It can be that `charging_end_time` only works (i.e. displays a timestamp in the UI) if you are already on 2021.12 (it was developed on the 2021.12 beta).

So if `charging_end_time_original` is close to the app (ignore 12h/24h), the new sensor should work with the new HA release.

**Comment by rhoddan** [comment]
Thanks for info. I will wait for 2021.12 and see if it works

**Comment by rhoddan** [comment]
Hi again, The released 2021.12 BMW version seems to work great! The only issue I can see is still the difference of remaining charge time 
![IMG_0164](https://user-images.githubusercontent.com/33499995/145899547-f9051ca2-09ae-4a7d-af59-92c12422308d.PNG)
![IMG_0165](https://user-images.githubusercontent.com/33499995/145899557-4717d148-1b6c-4695-ae17-2534e1900a8b.jpeg)

Thanks for a fantastic work!

**Comment by rikroe** [comment]
This should finally be fixed with Home Assistant 2021.12.4. 

Please uninstall the custom component again to be able to get updates with Home Assistant.

If new errors come up, please open new issues with the [home-assistant/core](https://github.com/home-assistant/core) or the [bimmerconnected/bimmer_connected](https://github.com/bimmerconnected/bimmer_connected) repositories.

**Comment by rhoddan** [comment]
Thanks - seems to work fine in 2021.12.4
